engine/runtime/src/main/scala/org/enso/interpreter/runtime/util/DiagnosticFormatter.scala

```diff
@@ -0,0 +1,216 @@
+          val isLocationEmpty = startColumn == endColumn
+          str ++= fansi
+            .Str(srcPath + ":" + lineNumber + ":" + startColumn + ": ")
+            .overlay(fansi.Bold.On)
+          str ++= fansi.Str(subject).overlay(textAttrs)
+          str ++= diagnostic.formattedMessage(fileLocationFromSection)
+          if (!isLocationEmpty) {
+            str ++= "\n"
+            str ++= oneLineFromSourceColored(lineNumber, startColumn, endColumn)
+            str ++= "\n"
+            str ++= underline(startColumn, endColumn)
+          }
+        } else {
+          str ++= fansi
+            .Str(
+              srcPath + ":[" + section.getStartLine + ":" + section.getStartColumn + "-" + section.getEndLine + ":" + section.getEndColumn + "]: "
+            )
+            .overlay(fansi.Bold.On)
+          str ++= fansi.Str(subject).overlay(textAttrs)
+          str ++= diagnostic.formattedMessage(fileLocationFromSection)
+          str ++= "\n"
+          val printAllSourceLines =
+            section.getEndLine - section.getStartLine <= maxSourceLinesToPrint
+          val endLine =
+            if (printAllSourceLines) section.getEndLine
+            else section.getStartLine + maxSourceLinesToPrint
+          for (lineNum <- section.getStartLine to endLine) {
+            str ++= oneLineFromSource(lineNum)
+            str ++= "\n"
+          }
+          if (!printAllSourceLines) {
+            val restLineCount =
+              section.getEndLine - section.getStartLine - maxSourceLinesToPrint
+            str ++= blankLinePrefix + "... and " + restLineCount + " more lines ..."
+            str ++= "\n"
+          }
+        }
+        if (outSupportsAnsiColors) {
+          str.render.stripLineEnd
+        } else {
+          str.plainText.stripLineEnd
+        }
+      case None =>
+        // There is no source section associated with the diagnostics
+        var str = fansi.Str()
+        val fileLocation = diagnostic.location match {
+          case Some(_) =>
+            fileLocationFromSectionOption(diagnostic.location, source)
+          case None =>
+            Option(source.getPath).getOrElse("<Unknown source>")
+        }
+
+        str ++= fansi
+          .Str(fileLocation)
+          .overlay(fansi.Bold.On)
+        str ++= ": "
+        str ++= fansi.Str(subject).overlay(textAttrs)
+        str ++= diagnostic.formattedMessage(fileLocationFromSection)
+        if (outSupportsAnsiColors) {
+          str.render.stripLineEnd
+        } else {
+          str.plainText.stripLineEnd
+        }
+    }
+  }
+
+  /** @see https://github.com/termstandard/colors/
+    * @see https://no-color.org/
+    * @return
+    */
+  private def outSupportsColors: Boolean = {
+    if (System.console() == null) {
+      // Non-interactive output is always without color support
+      return false
+    }
+    if (isOutputRedirected) {
+      return false
+    }
+    val ensoCtx = EnsoContext.get(null)
```

Comment:
I am not completely convinced myself, but adding `EnsoContext.isColorTerminalOutput()` might be safer (less open) choice than exposing all variables with `getEnvionmentVariable`.

Reply:
Refactored in 2f95a765ae88c357c592a494c755f54927aeb309